uniffi_bindgen/src/bindings/gecko/templates/SharedHeaderTemplate.h

```diff
@@ -0,0 +1,794 @@
+    });
+  }
+
+  int16_t ReadInt16() { return BitwiseCast<int16_t>(ReadUInt16()); }
+
+  uint32_t ReadUInt32() {
+    return ReadAt<uint32_t>([this](size_t aOffset) {
+      uint32_t value;
+      memcpy(&value, &mBuffer.mData[aOffset], sizeof(uint32_t));
+      return PR_ntohl(value);
+    });
+  }
+
+  int32_t ReadInt32() { return BitwiseCast<int32_t>(ReadUInt32()); }
+
+  uint64_t ReadUInt64() {
+    return ReadAt<uint64_t>([this](size_t aOffset) {
+      uint64_t value;
+      memcpy(&value, &mBuffer.mData[aOffset], sizeof(uint64_t));
+      return PR_ntohll(value);
+    });
+  }
+
+  int64_t ReadInt64() { return BitwiseCast<int64_t>(ReadUInt64()); }
+
+  float ReadFloat() { return BitwiseCast<float>(ReadUInt32()); }
+
+  double ReadDouble() { return BitwiseCast<double>(ReadUInt64()); }
+
+  /// Reads a length-prefixed UTF-8 encoded string at the current offset. The
+  /// closure takes a `Span` pointing to the raw bytes, which it can use to
+  /// copy the bytes into an `nsCString` or `nsString`.
+  ///
+  /// Safety: The closure must copy the span's contents into a new owned string.
+  /// It must not hold on to the span, as its contents will be invalidated when
+  /// the backing Rust byte buffer is freed. It must not call any other methods
+  /// on the reader.
+  template <typename T>
+  void ReadRawString(
+      T& aString,
+      const std::function<void(Span<const char>, T& aString)>& aClosure) {
+    uint32_t length = ReadInt32();
+    CheckedInt<size_t> newOffset = mOffset;
+    newOffset += length;
+    AssertInBounds(newOffset);
+    const char* begin =
+        reinterpret_cast<const char*>(&mBuffer.mData[mOffset.value()]);
+    aClosure(Span(begin, length), aString);
+    mOffset = newOffset;
+  }
+
+ private:
+  void AssertInBounds(const CheckedInt<size_t>& aNewOffset) const {
+    MOZ_RELEASE_ASSERT(aNewOffset.isValid() &&
+                       static_cast<int64_t>(aNewOffset.value()) <=
+                           mBuffer.mLen);
+  }
+
+  template <typename T>
+  T ReadAt(const std::function<T(size_t)>& aClosure) {
+    CheckedInt<size_t> newOffset = mOffset;
+    newOffset += sizeof(T);
+    AssertInBounds(newOffset);
+    T result = aClosure(mOffset.value());
+    mOffset = newOffset;
+    return result;
+  }
+
+  const RustBuffer& mBuffer;
+  CheckedInt<size_t> mOffset;
+};
+
+class MOZ_STACK_CLASS Writer final {
+ public:
+  explicit Writer(size_t aCapacity) : mBuffer(aCapacity) {}
+
+  void WriteUInt8(const uint8_t& aValue) {
+    WriteAt<uint8_t>(aValue, [this](size_t aOffset, const uint8_t& aValue) {
+      mBuffer[aOffset] = aValue;
+    });
+  }
+
+  void WriteInt8(const int8_t& aValue) {
+    WriteUInt8(BitwiseCast<uint8_t>(aValue));
+  }
+
+  // This code uses `memcpy` and other eye-twitchy patterns because it
+  // originally wrote values directly into a `RustBuffer`, instead of
+  // an intermediate `nsTArray`. Once #251 is fixed, we can return to
+  // doing that, and remove `ToRustBuffer`.
```

Comment:
The `Reader` code also uses `memcpy` but it doesn't have a comment like this. Should it?

Also, there should be a mention of big-endian encoding to match the one in `Reader`.

Reply:
I actually removed this comment because `Writer` writes directly into a `RustBuffer`, since it's now growable and we can avoid copying everything into the `nsTArray`, then copying everything again into a `RustBuffer`.